src/main/java/org/isf/opd/service/OpdIoOperationRepositoryImpl.java

```diff
@@ -61,15 +67,72 @@
 					getResultList();
 	}	
 
-	private TypedQuery<Opd> _getOpdQuery(
-			String diseaseTypeCode,
-			String diseaseCode, 
-			GregorianCalendar dateFrom,
-			GregorianCalendar dateTo,
-			int ageFrom, 
-			int ageTo,
-			char sex,
-			char newPatient) {
+    public List<Opd> findAllOpdWhereParamsWithPagination(
+            String diseaseTypeCode,
+            String diseaseCode,
+            GregorianCalendar dateFrom,
+            GregorianCalendar dateTo,
+            int ageFrom,
+            int ageTo,
+            char sex,
+            char newPatient, int pageNumber, int pageSize) {
+        TypedQuery<Opd> opdTypedQuery = getOpdIdQuery(diseaseTypeCode, diseaseCode, dateFrom, dateTo, ageFrom, ageTo, sex, newPatient);
+        opdTypedQuery.setFirstResult(pageNumber * pageSize);
+        opdTypedQuery.setMaxResults(pageSize);
+        return opdTypedQuery.getResultList();
+    }
+
+    public TypedQuery<Opd> getOpdIdQuery(
+            String diseaseTypeCode,
+            String diseaseCode,
+            GregorianCalendar dateFrom,
+            GregorianCalendar dateTo,
+            int ageFrom,
+            int ageTo,
+            char sex,
+            char newPatient) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Opd> criteriaQuery = criteriaBuilder.createQuery(Opd.class);
+        Metamodel m = entityManager.getMetamodel();
+        EntityType<Opd> opdEntityType = m.entity(Opd.class);
+        EntityType<Disease> diseaseEntityType = m.entity(Disease.class);
+
+        Root<Opd> opd = criteriaQuery.from(Opd.class);
+        criteriaQuery.select(opd);
+        Join<Opd, Disease> diseaseJoin = opd.join(opdEntityType.getSingularAttribute("disease", Disease.class));
+        Join<Disease, DiseaseType> diseaseTypeJoin = diseaseJoin.join(diseaseEntityType.getSingularAttribute("diseaseType", DiseaseType.class));
+        List<Predicate> predicates = new ArrayList<Predicate>();
+        if (!(diseaseTypeCode.equals(MessageBundle.getMessage("angal.opd.alltype")))) {
+            predicates.add(criteriaBuilder.equal(diseaseTypeJoin.get("code"), diseaseTypeCode));
+        }
+        if (!diseaseCode.equals(MessageBundle.getMessage("angal.opd.alldisease"))) {
+            predicates.add(criteriaBuilder.equal(diseaseJoin.get("code"), diseaseCode));
+        }
+        if (ageFrom != 0 || ageTo != 0) {
+			predicates.add(criteriaBuilder.between(opd.get("age"), ageFrom, ageTo));
+		}
+        if (sex != 'A') {
+            predicates.add(criteriaBuilder.equal(opd.get("sex"), sex));
+        }
+        if (newPatient != 'A') {
+            predicates.add(criteriaBuilder.equal(opd.get("newPatient"), newPatient));
+        }
+		predicates.add(criteriaBuilder.between(opd.get("visitDate"), dateFrom, dateTo));
+
+		criteriaQuery.where(predicates.toArray(new Predicate[0]));
+
+        return entityManager.createQuery(criteriaQuery);
```

Comment:
The formatting in this section (lines 111-124) seems a bit off as in other locations in the file; maybe just reformat the whole file?

Reply:
Done, thanks!